diff --git a/agent/call-pinentry.c b/agent/call-pinentry.c
index c6c52be..ac6e6e9 100644
--- a/agent/call-pinentry.c
+++ b/agent/call-pinentry.c
@@ -129,7 +129,7 @@ void
 agent_query_dump_state (void)
 {
   log_info ("agent_query_dump_state: entry_ctx=%p pid=%ld popup_tid=%p\n",
-            entry_ctx, (long)assuan_get_pid (entry_ctx), (void*)popup_tid);
+            entry_ctx, (long)assuan_get_pid (entry_ctx), (void*)popup_tid.__);
 }
 
 /* Called to make sure that a popup window owned by the current
@@ -137,7 +137,7 @@ agent_query_dump_state (void)
 void
 agent_reset_query (ctrl_t ctrl)
 {
-  if (entry_ctx && popup_tid && ctrl->pinentry_active)
+  if (entry_ctx && popup_tid.__  && ctrl->pinentry_active)
     {
       agent_popup_message_stop (ctrl);
     }
@@ -2107,7 +2107,7 @@ agent_popup_message_stop (ctrl_t ctrl)
   if (ctrl->pinentry_mode == PINENTRY_MODE_LOOPBACK)
     return;
 
-  if (!popup_tid || !entry_ctx)
+  if (!popup_tid.__ || !entry_ctx)
     {
       log_debug ("agent_popup_message_stop called with no active popup\n");
       return;
