diff --git a/configure b/configure
index aca916f..4ae9025 100755
--- a/configure
+++ b/configure
@@ -9207,7 +9207,7 @@ $as_echo "$as_me: WARNING:
 # FiXME: Use GNUPG_CHECK_LIBUSB and modify to use separate AC_SUBSTs.
 if test "$use_ccid_driver" = auto || test "$use_ccid_driver" = yes; then
    case "${host}" in
-     *-mingw32*)
+     *-mingw32* | *-*-openedition)
        LIBUSB_NAME=
        LIBUSB_LIBS=
        LIBUSB_CPPFLAGS=
