diff --git a/tests/openpgp/encryptp.scm b/tests/openpgp/encryptp.scm
index 0f09a1e..c61cd37 100755
--- a/tests/openpgp/encryptp.scm
+++ b/tests/openpgp/encryptp.scm
@@ -26,7 +26,7 @@
    (tr:do
     (tr:open source)
     (tr:pipe-do
-     (pipe:gpg `(--yes --encrypt --recipient ,usrname2))
+     (pipe:gpg `(--yes --encrypt --recipient ,usrname2 --trust-model always))
      (pipe:gpg '(--yes --decrypt)))
     (tr:assert-identity source)))
  (append plain-files data-files))
