diff --git a/g10/decrypt.c b/g10/decrypt.c
index cb9e36a..9314d23 100644
--- a/g10/decrypt.c
+++ b/g10/decrypt.c
@@ -53,7 +53,9 @@ decrypt_message (ctrl_t ctrl, const char *filename)
 
   /* Open the message file.  */
   fp = iobuf_open (filename);
-  if (fp && is_secured_file (iobuf_get_fd (fp)))
+  int fd =iobuf_get_fd (fp);
+  __disableautocvt(fd);
+  if (fp && is_secured_file (fd))
     {
       iobuf_close (fp);
       fp = NULL;
